Dev/field composition

kettanaito opened the pull request:
Implements #132. This pull request introduces a breaking change.

## Changes
1. Generic `Field` class was deleted in favor of field composition vs inheritance
1. A new `createField` HOC was introduced
1. Each native instance of fields was re-written to use the mentioned HOC
1. Library size decreased by 7 Kb

## Needs discussion
This is how you would create custom fields now:

```jsx
import React from 'react';
import { createField } from 'react-advanced-form';
import ThirdPartyField from '...';

class CustomField extends React.Component {
  render() {
    return (<ThirdPartyField { ...this.props } />
  }
}

export default createField({
  valueProp: 'checked', // update "checked" prop onChange
  mapPropsToField: props => ({
    value: props.checked ? props.value : null // custom props remapping for field registration
  }),
  enforceProps: (props, contextProps) => ({
    customProp: (props.some === contextProps.some)
  })
})(CustomField);
```

I would like to discuss the API mainly.

### Method names
* `valueProp: string` (default: `"value"`) a prop name of the field which should be treated as `value` during the onChange updates. **Note:** some fields (i.e. Radio) have both `value` and `valueProp: 'checked'`. That means `checked` is getting updated during onChange, but `value` is still important for the field's serialization value.
* `mapPropsToField(props: Object): Object` An optional method which returns the props to be treated as registration props during the field's registration. Useful for custom logic if props (i.e Radio field).
* `enforceProps(props: Object, contextProps: Map): Object` An optional method to provide custom props to the field component based on the existing props and contextProps. Usecase: `Field.Radio` should always take the `value` from its direct props, and never from the `contextProps`. Its `checked` property is handled in the similar way.

Please give me your feedback if those method names are clear enough, and any suggestions you feel are necessary. Thanks.

---

My personal thoughts on the methods are that there are very non intuitive things going on, for example after `enforceProps`. The props returned from `enforceProps` are assigned to the `WrappedComponent` and are accessible there through `this.props`, as usual.

I also don't think `mapPropsToField` is descriptive enough. This method should state that the props you return will be used as the props during the field registration. I like the beginning - `mapPropsTo`, but I am confused what is the descriptive target to call.

kettanaito commented:
Also, #134 should be merged **before** this change, if possible.

Although its implementation may change due to the breaking change of this pull request, it's bad to have `0.15.1` with the bug of this kind of severity. So we should fix it for both versions.

kettanaito commented:
1. How to set custom event handlers (onChange)?

Vidlec commented on src/Fields/Checkbox.jsx:
```diff
@@ -1,62 +1,18 @@
+export default createField({
+  valueProp: 'checked',
+  mapPropsToField: props => ({
+    ...props,
```
Is passing `...props` there needed?
I thought `return (<input { ...props } />);` Does the job?

kettanaito replied:
When you provide `mapPropsToField` you are expected to return *all* the refistration props, including default props.

I wouldn't add any default extension logic, this would only complicate things.

kettanaito replied:
What you may be getting wrong is where each props go.

`mapPropsToField` creates the record object in 'Form.state.field`.
`<input { ...props } />` destructs the composed props passed from the `connectField`'s `Field` class to the custom WrappedComponent.
`enforceProps` appends explicit props to the logic above.

We cannot access WrappedComponent.props in createField and Field class. The props can flow only down from Field to WrappedComponent. Form.state.field record is not related to props at all, that's a separate required object for field record management.

Vidlec commented on src/createField.jsx:
```diff
@@ -0,0 +1,256 @@
+      }
+
+      /**
+       * Ensures "this.contextProps" is updated accoding to the component updates.
```
Literally unreadable. 
JK - just a typo... :)

kettanaito replied:
How could I...

Vidlec commented on src/utils/fieldUtils/serializeFields.js:
```diff
@@ -13,7 +13,8 @@ export default function serializeFields(fields, valueResolver = null) {
     if (!isCheckbox && hasEmptyValue) return serialized;
 
-    const defaultValue = isCheckbox ? fieldProps.get('checked') : fieldProps.get('value');
+    const valueProp = fieldProps.get('valueProp');
```
`valuePropName`?

left me confused for a sec

kettanaito replied:
I like the idea to rename to `valuePropName`.

Vidlec commented:
This overall looks VERY good. 
Name `mapPropsToField` sounds good to me, I would leave it as it is :)
On the other hand `valueProp` sounds weird... `valueName`, `valuePropName`, `valueProperty (yes, its essentially valueProp :))`?
 I dont know. Naming is hard

kettanaito commented:
* Fixed comment typo
* Renamed `valueProp` to `valuePropName`

Still need to think of custom event handlers and how to provide those. Using HOC options for this purpose seems redundant, as it's out of the component's scope at the time of the declaration.

@Vidlec, any ideas are appreciated.

kettanaito commented:
Updated the documentation to reflect the latest changes.

Need to check the approach for custom event handlers mentioned in #132.

kettanaito commented:
@Vidlec do you have any other comments regarding this change?

If not, I would aim toward merging it this week. A few tests missing, but overall it should be okay to rollout in a breaking release.
